src/sandstorm/grain.capnp

```diff
@@ -126,6 +126,99 @@ interface SandstormApi(AppObjectId) {
   # This method is here on `SandstormApi` rather than on `Identity` in order to ease a possible
   # future transition to a model where identity IDs are not globally stable and each grain has a
   # separate (possibly incompatible) map from identity ID to account ID.
+
+  schedule @11 ScheduledJob;
+  # Schedule a callback to be executed in the future.
+  #
+  # The job may be scheduled to execute once at a specific time, or periodically. See
+  # `ScheduledJob` for more information.
+  #
+  # When this method returns, the job will have been scheduled.
+}
+
+struct ScheduledJob {
+  # A job to be scheduled by `SandstormApi.schedule()`.
+
+  name @0 :Util.LocalizedText;
+  # A name for the job. This will be used to describe the job to the user.
+
+  callback @1 :Callback;
+  # A callback to actually execute when it is time to run the job.
+  #
+  # In addition to implementing the `Callback` interface below, this must also
+  # implement `AppPersistent`. When a request to schedule the job is made, Sandstorm
+  # will `save()` the callback, and then later `restore()` it when it is time to
+  # run the job. See `AppPersistent` for more details on how this works.
+  #
+  # Note that, in some circumstances, the app may need to save some information
+  # locally to recognize the task. There is a possibility that between scheduling
+  # the job and actually saving the information, The app may be killed, the server
+  # may crash, or some other failure may occur that prevents the app from saving
+  # the information. We recommend dealing with this in one of the following ways:
+  #
+  # - Where possible, save all needed information in the `objectId` returned
+  #   by `AppPersistent.save()`; this avoids the race condition entirely.
+  # - In your implementation of `MainView.restore()`, be sure to correctly
+  #   handle attempts to restore unrecognized capabilities, by raising an
+  #   exception with `type = failed`. This will cause Sandstorm to cancel
```

Comment:
I think cancelling the schedule if it throws an exception would lead to scheduling being unreliable.

Exceptions may be thrown for transient reasons. A "failed" exception in KJ means "there is a bug in the code", but the nature of the bug could be just about anything. It could be a failure to allocate memory, or a race condition. It could be that the grain is temporarily in a transitional state due to poorly-written storage code. It could be that the grain was updated to a new version of the app which has a bug, but an urgent bugfix update is on its way.

Generally, "reliable" message delivery systems have to retry on error, with some sort of back-off to avoid spinning and burning resources. If retries keep failing, a human should be alerted to the problem, but the system cannot give up without a human telling it to.

For example, consider Stripe webhooks. Stripe sends an HTTP request to your server to notify you of some event, like a failed payment. If your server returns a 4xx or 5xx error, Stripe retries later. If requests keep failing, Stripe will actually send the developer an e-mail to say "hey go fix your server so we can deliver these events". Stripe will never let an event go undelivered. This has saved my ass many times! If not for this behavior, every time we deployed a buggy webhook handler I would have been stuck going back through Stripe events manually to figure out what was missed.

Reply:
Yeah, I see your point. Spitballing: what if, instead of using exceptions, we modified the return type of `restore` to allow a "not-recognized" response? This would have the same advantages to the existing proposal, but without the possibility of confusing it with a different type of error.

Reply:
Well, if the app explicitly knows that this is an event which should be un-scheduled, it could always construct a callback that returns false.

My worry about having a "not found" error code is that I've seen too many real-world examples where buggy or misconfigured applications incorrectly returned "not found".

Reply:
Having the app set `cancelFutureRuns = true` sounds reasonable to me.

Reply:
Hm, but the trouble with that is when calling `restore` you don't know how the returned capability is going to be used, so choosing to return something that responds to `Callback.run` is a bit awkward in general.

More spitballing: would unscheduling if run raises `unimplemented` be better than `restore` raising `failed`? I still worry about something bubbling up from a nested call, my gut is that's not a great idea.

Reply:
If the app knows that `restore()` is trying to restore a cron callback, then it can return a dummy callback that returns false.

If the app does not know what kind of object is being restored, then it seems like fundamentally the app cannot signal explicitly that the schedule should be canceled.

FWIW since apps are allowed to define their own SturdyRef type, the app can encode in the ref what type of ref it is, so can potentially know this way when to return a dummy callback without having to store anything... but then again we're back to the problem where the app is pretty fragile, and a bad database query that happens to return no results can mean the job gets unscheduled when it shouldn't have.

I really think the right answer is to alert the user that this background job has stopped working. If the user wants to cancel the job, they can do so, but they can also try to debug the problem.